src/snowflake/snowpark/ml/transformer.py

```diff
@@ -0,0 +1,395 @@
+            self._category_table_name.items(),
+        ):
+            category_table = df._session.table(table_name)
+            # if sparse is set to true, covert category_table so that it store onehot array
+            if sparse:
+                df_sparse_array = category_table.with_column("sparse", lit(0)).select(
+                    [array_agg(col("sparse")).as_(f"{fc}_sparse")]
+                )
+                df_sparse_array = category_table.join(df_sparse_array).join(
+                    df_count_table
+                )
+                category_table = df_sparse_array.select(
+                    [
+                        col(fc),
+                        array_slice(
+                            array_insert(col(f"{fc}_sparse"), col(ec), lit(1)),
+                            lit(0),
+                            col(f"{fc}_count"),
+                        ).as_(ec),
+                    ]
+                )
+                df = category_table.join(
+                    df, df[c] == category_table[fc], join_type="right"
+                ).join(df_sparse_array.select([col(f"{fc}_sparse")]).limit(1))
+            else:
+                df = category_table.join(
+                    df, df[c] == category_table[fc], join_type="right"
+                )
+            if self._merge:
+                if not sparse:
+
+                    encoder_structure.append(
+                        iff(col(ec).is_null(), col(ec), col(ec) + encoder_offset)
+                    )
+                    encoder_offset += col(f"{fc}_count")
+                else:
+                    encoder_structure = array_cat(
+                        encoder_structure,
+                        iff(col(ec).is_null(), df[f"{fc}_sparse"], col(ec)),
+                    )
+        # depending on self._merge and self.sparse, prepare the final output
+        if self._merge:
+            if not sparse:
+
+                encoder_structure = array_construct_compact(*encoder_structure)
+                onehot_array = array_construct(
+                    df["encoder_count"], encoder_structure, lit(1)
+                )
+            else:
+                onehot_array = encoder_structure
+        if self._merge:
+            original_columns.append(self._output_cols[0])
+            df = df.with_column(self._output_cols[0], onehot_array)
+            df = df.select([col(c) for c in original_columns])
+        else:
+            for fc, oc in zip(self._states_table_cols, self._output_cols):
+                if not sparse:
+                    onehot_array = array_construct_compact(
+                        df[f"{fc}_count"], df[f"{fc}_encoder"], lit(1)
+                    )
+                else:
+                    onehot_array = iff(
+                        df[f"{fc}_encoder"].is_null(),
+                        df[f"{fc}_sparse"],
+                        df[f"{fc}_encoder"],
+                    )
+                df = df.with_column(oc, onehot_array)
```

Comment:
Could you use [with_columns](https://github.com/snowflakedb/snowpark-python/blob/e0cc59f81bc48e027223528d112825edc9dd0b65/src/snowflake/snowpark/dataframe.py#L1764) here to add multiple columns? It will optimize the query under the hood. The same for ordinal encoder and binarizer.